stellargraph/mapper/node_mappers.py

```diff
@@ -1128,3 +1128,275 @@ def flow_from_dataframe(self, node_targets, shuffle=False):
+            raise TypeError("Graph must be a StellarGraph object.")
+
+        self.graph = G
+        self.name = name
+        self.use_sparse = sparse
+
+        # Check if the graph has features
+        G.check_graph_for_ml()
+
+        # extract node, feature, and edge type info from G
+        self.node_list = list(G.nodes())
+
+        self.features = G.get_feature_for_nodes(self.node_list)
+
+        edge_types = sorted(set(e[-1] for e in G.edges))
+
+        node_index = dict(zip(self.node_list, range(len(self.node_list))))
+
+        # create a list of adjacency matrices - one adj matrix for each edge type
+        # an adjacency matrix is created for each edge type from all edges of that type
+        self.As = []
+
+        for edge_type in edge_types:
+
+            col_index = [
+                node_index[n1] for n1, n2, etype in G.edges if etype == edge_type
+            ]
+            row_index = [
+                node_index[n2] for n1, n2, etype in G.edges if etype == edge_type
+            ]
+            data = np.ones(len(col_index), np.float64)
+
+            # note that A is the transpose of the standard adjacency matrix
+            # this is to aggregate features from incoming nodes
+            A = sps.coo_matrix(
+                (data, (row_index, col_index)),
+                shape=(len(self.node_list), len(self.node_list)),
+            )
+
+            if transform is None:
+                A = normalize_adj(A, symmetric=False)
+            else:
+                self.features, A = transform(self.features, A)
+
+            A = A.tocoo()
+            self.As.append(A)
+
+        # Get the features for the nodes
+        self.features = G.get_feature_for_nodes(self.node_list)
+
+    def flow(self, node_ids, targets=None):
+        """
+        Creates a generator/sequence object for training or evaluation
+        with the supplied node ids and numeric targets.
+
+        Args:
+            node_ids: and iterable of node ids for the nodes of interest
+                (e.g., training, validation, or test set nodes)
+            targets: a 2D array of numeric node targets with shape `(len(node_ids), target_size)`
+
+        Returns:
+            A NodeSequence object to use with RGCN models
+            in Keras methods :meth:`fit_generator`, :meth:`evaluate_generator`,
+            and :meth:`predict_generator`
+        """
+
+        if targets is not None:
+            # Check targets is an iterable
+            if not is_real_iterable(targets):
+                raise TypeError("Targets must be an iterable or None")
+
+            # Check targets correct shape
+            if len(targets) != len(node_ids):
+                raise TypeError("Targets must be the same length as node_ids")
+
+        # The list of indices of the target nodes in self.node_list
+        # use dictionary for faster index look-up time
+        node_index = dict(zip(self.node_list, range(len(self.node_list))))
```

Comment:
You perform the same operation in the `__init__` method. Can you do it once and store it with the object in `self.node_index`? Or are you trying to save memory by using a temporary object?

Reply:
I was trying to save memory, but I don't think time/memory overhead will be very much either. What do you think?

Reply:
Let's just do it in __init__. It will be easier to move it out of __init__ later, especially after we have re-worked the StellarGraph class and we have a better memory management solution.

Reply:
done!